src/dvc_data/index/diff.py

```diff
@@ -246,44 +247,56 @@ def _detect_renames(changes: Iterable[Change]):
 
     for change in changes:
         if change.typ == ADD:
+            assert change.new
             added.append(change)
         elif change.typ == DELETE:
+            assert change.old
             deleted.append(change)
         else:
             yield change
 
     def _get_key(change):
         return change.key
 
-    added[:] = sorted(added, key=_get_key)
-    deleted[:] = sorted(deleted, key=_get_key)
+    # Sort the lists to maintain the same order
+    # as older implementation.
+    added.sort(key=_get_key)
+    deleted.sort(key=_get_key)
+
+    # Create a dictionary for fast lookup of deletions by hash_info
+    deleted_dict: dict[Optional[HashInfo], deque[Change]] = defaultdict(deque)
+    for change in deleted:
+        # We checked change.old for all deleted above, so cast
+        change_hash = cast(DataIndexEntry, change.old).hash_info
+        # appendleft to get queue behaviour (we pop off right)
+        deleted_dict[change_hash].appendleft(change)
```

Comment:
Instead of using `assert` and `cast`, let's keep change with `None` `hash_info`s.

```python
for change in deleted:
    hi = change.old.hash_info if change.old else None
    deleted_dict[hi].appendleft(change)
```


And then, inside loop, we can yield `None` values without any changes.

```python
for change in added:
  new_entry = change.new
  hi = new_entry.hash_info if new_entry else None
  if hi and (lst := deleted_dict.get(hi)):
      deletion = lst.pop()
      yield ...
  else:
      yield change
```

Reply:
I like this, nice!